File: snippets/screens/compose/search_results_screen/2/search_results_screen.kt
package com.example.compose.search

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.*
import androidx.compose.foundation.lazy.grid.*
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import kotlinx.coroutines.delay
import java.text.NumberFormat
import java.util.*

data class SearchResult(
    val id: String,
    val title: String,
    val description: String,
    val imageUrl: String?,
    val price: Double?,
    val rating: Double?,
    val reviewCount: Int?,
    val category: String?,
    val tags: List<String>,
    val dateAdded: Date?,
    val author: String?,
    val source: String?
)

data class ActiveFilter(
    val key: String,
    val value: String
)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SearchResultsScreen(
    initialQuery: String = "",
    initialFilters: Map<String, Any> = emptyMap(),
    onNavigateBack: () -> Unit = {}
) {
    var searchText by remember { mutableStateOf(initialQuery) }
    var results by remember { mutableStateOf<List<SearchResult>>(emptyList()) }
    var isLoading by remember { mutableStateOf(false) }
    var isLoadingMore by remember { mutableStateOf(false) }
    var hasMoreResults by remember { mutableStateOf(true) }
    var currentFilters by remember { mutableStateOf(initialFilters) }
    var sortBy by remember { mutableStateOf("relevance") }
    var isGridView by remember { mutableStateOf(false) }
    var showFilterSheet by remember { mutableStateOf(false) }
    var showSortSheet by remember { mutableStateOf(false) }

    var totalResults by remember { mutableStateOf(0) }
    var currentPage by remember { mutableStateOf(1) }
    val pageSize = 20

    val context = LocalContext.current

    fun performSearch(isNewSearch: Boolean = true) {
        if (isNewSearch) {
            results = emptyList()
            currentPage = 1
            isLoading = true
            hasMoreResults = true
        } else {
            isLoadingMore = true
            currentPage++
        }

        // Simulate API call
        LaunchedEffect(searchText + sortBy + currentFilters.toString() + currentPage) {
            delay(800)
            val newResults = generateMockResults(searchText, currentPage, pageSize)
            if (isNewSearch) {
                results = newResults
                totalResults = 157 // Mock total
            } else {
                results = results + newResults
            }
            isLoading = false
            isLoadingMore = false
            hasMoreResults = results.size < totalResults
        }
    }

    // Perform initial search
    LaunchedEffect(Unit) {
        if (searchText.isNotEmpty()) {
            performSearch()
        }
    }

    val activeFilters = remember(currentFilters) {
        currentFilters.entries.mapNotNull { (key, value) ->
            when (value) {
                is String -> if (value.isNotEmpty()) listOf(ActiveFilter(key, value)) else emptyList()
                is List<*> -> value.filterIsInstance<String>().map { ActiveFilter(key, it) }
                else -> emptyList()
            }
        }.flatten()
    }

    Column(
        modifier = Modifier.fillMaxSize()
    ) {
        // Search Bar
        TopAppBar(
            title = {
                OutlinedTextField(
                    value = searchText,
                    onValueChange = { searchText = it },
                    placeholder = { Text("Search...") },
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(imeAction = ImeAction.Search),
                    keyboardActions = KeyboardActions(onSearch = { performSearch() }),
                    modifier = Modifier.fillMaxWidth()
                )
            },
            navigationIcon = {
                IconButton(onClick = onNavigateBack) {
                    Icon(Icons.Default.ArrowBack, contentDescription = "Back")
                }
            },
            actions = {
                IconButton(onClick = { isGridView = !isGridView }) {
                    Icon(
                        if (isGridView) Icons.Default.ViewList else Icons.Default.GridView,
                        contentDescription = "Toggle view"
                    )
                }
                IconButton(onClick = { showFilterSheet = true }) {
                    Icon(Icons.Default.Tune, contentDescription = "Filters")
                }
            }
        )

        // Search Summary and Controls
        Surface(
            color = MaterialTheme.colorScheme.surfaceVariant,
            modifier = Modifier.fillMaxWidth()
        ) {
            Column(
                modifier = Modifier.padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = if (isLoading && results.isEmpty())
                            "Searching..."
                        else
                            "About ${NumberFormat.getNumberInstance().format(totalResults)} results for \"$searchText\"",
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )

                    TextButton(onClick = { showSortSheet = true }) {
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(4.dp)
                        ) {
                            Icon(
                                Icons.Default.Sort,
                                contentDescription = null,
                                modifier = Modifier.size(16.dp)
                            )
                            Text(sortBy.uppercase())
                        }
                    }
                }

                // Active Filters
                if (activeFilters.isNotEmpty()) {
                    LazyRow(
                        horizontalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        items(activeFilters) { filter ->
                            FilterChip(
                                onClick = { /* Remove filter */ },
                                label = {
                                    Row(
                                        verticalAlignment = Alignment.CenterVertically,
                                        horizontalArrangement = Arrangement.spacedBy(4.dp)
                                    ) {
                                        Text("${filter.key}: ${filter.value}")
                                        Icon(
                                            Icons.Default.Close,
                                            contentDescription = "Remove filter",
                                            modifier = Modifier.size(16.dp)
                                        )
                                    }
                                },
                                selected = true
                            )
                        }
                    }
                }
            }
        }

        // Results
        Box(modifier = Modifier.weight(1f)) {
            when {
                isLoading && results.isEmpty() -> {
                    Box(
                        modifier = Modifier.fillMaxSize(),
                        contentAlignment = Alignment.Center
                    ) {
                        CircularProgressIndicator()
                    }
                }

                results.isEmpty() -> {
                    EmptyResultsState(
                        onClearFilters = {
                            currentFilters = emptyMap()
                            performSearch()
                        }
                    )
                }

                isGridView -> {
                    ResultsGridView(
                        results = results,
                        searchQuery = searchText,
                        hasMoreResults = hasMoreResults,
                        isLoadingMore = isLoadingMore,
                        onLoadMore = { performSearch(false) }
                    )
                }

                else -> {
                    ResultsListView(
                        results = results,
                        searchQuery = searchText,
                        hasMoreResults = hasMoreResults,
                        isLoadingMore = isLoadingMore,
                        onLoadMore = { performSearch(false) }
                    )
                }
            }
        }
    }

    // Filter Sheet
    if (showFilterSheet) {
        ModalBottomSheet(
            onDismissRequest = { showFilterSheet = false }
        ) {
            FilterScreen(
                initialFilters = FilterState(), // Convert currentFilters to FilterState
                onApplyFilters = { filterState ->
                    // Convert FilterState to Map and update currentFilters
                    showFilterSheet = false
                    performSearch()
                },
                onDismiss = { showFilterSheet = false }
            )
        }
    }

    // Sort Sheet
    if (showSortSheet) {
        ModalBottomSheet(
            onDismissRequest = { showSortSheet = false }
        ) {
            SortScreen(
                currentSort = SortConfiguration(SortOption.RELEVANCE, true), // Convert sortBy
                onApplySort = { config ->
                    sortBy = config.option.key
                    showSortSheet = false
                    performSearch()
                },
                onNavigateBack = { showSortSheet = false }
            )
        }
    }
}

@Composable
private fun EmptyResultsState(
    onClearFilters: () -> Unit
) {
    Column(
        modifier = Modifier.fillMaxSize(),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Icon(
            imageVector = Icons.Default.SearchOff,
            contentDescription = null,
            modifier = Modifier.size(64.dp),
            tint = MaterialTheme.colorScheme.outline
        )

        Spacer(modifier = Modifier.height(16.dp))

        Text(
            text = "No results found",
            style = MaterialTheme.typography.titleLarge
        )

        Spacer(modifier = Modifier.height(8.dp))

        Text(
            text = "Try adjusting your search or filters",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.outline
        )

        Spacer(modifier = Modifier.height(24.dp))

        Button(onClick = onClearFilters) {
            Text("Clear Filters")
        }
    }
}

@Composable
private fun ResultsListView(
    results: List<SearchResult>,
    searchQuery: String,
    hasMoreResults: Boolean,
    isLoadingMore: Boolean,
    onLoadMore: () -> Unit
) {
    LazyColumn(
        contentPadding = PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        items(results) { result ->
            ResultListItem(result, searchQuery)
        }

        if (hasMoreResults) {
            item {
                Box(
                    modifier = Modifier.fillMaxWidth(),
                    contentAlignment = Alignment.Center
                ) {
                    if (isLoadingMore) {
                        CircularProgressIndicator()
                    } else {
                        LaunchedEffect(Unit) { onLoadMore() }
                    }
                }
            }
        }
    }
}

@Composable
private fun ResultsGridView(
    results: List<SearchResult>,
    searchQuery: String,
    hasMoreResults: Boolean,
    isLoadingMore: Boolean,
    onLoadMore: () -> Unit
) {
    LazyVerticalGrid(
        columns = GridCells.Fixed(2),
        contentPadding = PaddingValues(16.dp),
        horizontalArrangement = Arrangement.spacedBy(16.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        items(results) { result ->
            ResultGridItem(result, searchQuery)
        }

        if (hasMoreResults) {
            item(span = { GridItemSpan(2) }) {
                Box(
                    modifier = Modifier.fillMaxWidth(),
                    contentAlignment = Alignment.Center
                ) {
                    if (isLoadingMore) {
                        CircularProgressIndicator()
                    } else {
                        LaunchedEffect(Unit) { onLoadMore() }
                    }
                }
            }
        }
    }
}

@Composable
private fun ResultListItem(
    result: SearchResult,
    searchQuery: String
) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        onClick = { /* Navigate to detail */ }
    ) {
        Row(
            modifier = Modifier.padding(16.dp),
            horizontalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            // Image
            AsyncImage(
                model = result.imageUrl,
                contentDescription = null,
                modifier = Modifier
                    .size(80.dp)
                    .clip(MaterialTheme.shapes.medium),
                contentScale = ContentScale.Crop,
                placeholder = {
                    Box(
                        modifier = Modifier
                            .size(80.dp)
                            .clip(MaterialTheme.shapes.medium),
                        contentAlignment = Alignment.Center
                    ) {
                        Icon(
                            Icons.Default.Image,
                            contentDescription = null,
                            tint = MaterialTheme.colorScheme.outline
                        )
                    }
                }
            )

            Column(
                modifier = Modifier.weight(1f),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                // Title
                Text(
                    text = highlightText(result.title, searchQuery),
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis
                )

                // Description
                Text(
                    text = result.description,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    maxLines = 3,
                    overflow = TextOverflow.Ellipsis
                )

                // Metadata Row
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    if (result.rating != null) {
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(4.dp)
                        ) {
                            Icon(
                                Icons.Default.Star,
                                contentDescription = null,
                                tint = Color(0xFFFFB000),
                                modifier = Modifier.size(16.dp)
                            )
                            Text(
                                text = String.format("%.1f", result.rating),
                                style = MaterialTheme.typography.bodySmall
                            )
                        }
                    }

                    if (result.category != null) {
                        AssistChip(
                            onClick = { },
                            label = { Text(result.category) }
                        )
                    }

                    Spacer(modifier = Modifier.weight(1f))

                    if (result.price != null) {
                        Text(
                            text = "$${String.format("%.2f", result.price)}",
                            style = MaterialTheme.typography.titleMedium,
                            fontWeight = FontWeight.Bold,
                            color = MaterialTheme.colorScheme.primary
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun ResultGridItem(
    result: SearchResult,
    searchQuery: String
) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        onClick = { /* Navigate to detail */ }
    ) {
        Column {
            // Image
            AsyncImage(
                model = result.imageUrl,
                contentDescription = null,
                modifier = Modifier
                    .fillMaxWidth()
                    .height(120.dp),
                contentScale = ContentScale.Crop,
                placeholder = {
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(120.dp),
                        contentAlignment = Alignment.Center
                    ) {
                        Icon(
                            Icons.Default.Image,
                            contentDescription = null,
                            tint = MaterialTheme.colorScheme.outline
                        )
                    }
                }
            )

            Column(
                modifier = Modifier.padding(12.dp),
                verticalArrangement = Arrangement.spacedBy(4.dp)
            ) {
                Text(
                    text = highlightText(result.title, searchQuery),
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.Bold,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis
                )

                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    if (result.rating != null) {
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(2.dp)
                        ) {
                            Icon(
                                Icons.Default.Star,
                                contentDescription = null,
                                tint = Color(0xFFFFB000),
                                modifier = Modifier.size(14.dp)
                            )
                            Text(
                                text = String.format("%.1f", result.rating),
                                style = MaterialTheme.typography.bodySmall
                            )
                        }
                    }

                    if (result.price != null) {
                        Text(
                            text = "$${result.price.toInt()}",
                            style = MaterialTheme.typography.titleSmall,
                            fontWeight = FontWeight.Bold,
                            color = MaterialTheme.colorScheme.primary
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun highlightText(text: String, query: String) = buildAnnotatedString {
    if (query.isEmpty()) {
        append(text)
        return@buildAnnotatedString
    }

    val lowercaseText = text.lowercase()
    val lowercaseQuery = query.lowercase()
    var startIndex = 0

    while (true) {
        val index = lowercaseText.indexOf(lowercaseQuery, startIndex)
        if (index == -1) {
            append(text.substring(startIndex))
            break
        }

        append(text.substring(startIndex, index))
        withStyle(
            style = SpanStyle(
                color = MaterialTheme.colorScheme.primary,
                fontWeight = FontWeight.Bold
            )
        ) {
            append(text.substring(index, index + query.length))
        }
        startIndex = index + query.length
    }
}

private fun generateMockResults(query: String, page: Int, pageSize: Int): List<SearchResult> {
    return (1..pageSize).map { index ->
        val baseIndex = (page - 1) * pageSize + index
        SearchResult(
            id = "result_$baseIndex",
            title = "Search Result $baseIndex - $query",
            description = "This is a detailed description for search result $baseIndex. It contains relevant information about $query and provides useful context for the user.",
            imageUrl = "https://picsum.photos/300/200?random=$baseIndex",
            price = if (baseIndex % 5 == 0) 29.99 + (baseIndex * 2.5) else null,
            rating = 3.0 + (baseIndex % 3),
            reviewCount = 50 + (baseIndex * 12),
            category = listOf("Electronics", "Books", "Clothing", "Home")[baseIndex % 4],
            tags = listOf("Popular", "New", "Featured"),
            dateAdded = Date(System.currentTimeMillis() - (baseIndex * 86400000L)),
            author = "Author ${baseIndex % 10}",
            source = "Source ${baseIndex % 5}"
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun SearchResultsScreenPreview() {
    MaterialTheme {
        SearchResultsScreen(initialQuery = "Kotlin programming")
    }
}

File: snippets/screens/swiftui/search_results_screen/2/search_results_screen.swift
import SwiftUI

struct SearchResult {
    let id: String
    let title: String
    let description: String
    let imageUrl: String?
    let price: Double?
    let rating: Double?
    let reviewCount: Int?
    let category: String?
    let tags: [String]
    let dateAdded: Date?
    let author: String?
    let source: String?
}

struct SearchResultsScreen: View {
    let query: String?
    let filters: [String: Any]?

    @State private var searchText: String
    @State private var results: [SearchResult] = []
    @State private var isLoading = false
    @State private var isLoadingMore = false
    @State private var hasMoreResults = true
    @State private var currentFilters: [String: Any] = [:]
    @State private var sortBy = "relevance"
    @State private var isGridView = false
    @State private var showFilterSheet = false
    @State private var showSortSheet = false

    @State private var totalResults = 0
    @State private var currentPage = 1
    private let pageSize = 20

    init(query: String? = nil, filters: [String: Any]? = nil) {
        self.query = query
        self.filters = filters
        self._searchText = State(initialValue: query ?? "")
        self._currentFilters = State(initialValue: filters ?? [:])
    }

    var body: some View {
        NavigationView {
            VStack(spacing: 0) {
                // Search Bar
                HStack {
                    TextField("Search...", text: $searchText)
                        .textFieldStyle(.roundedBorder)
                        .onSubmit {
                            performSearch()
                        }

                    Button {
                        isGridView.toggle()
                    } label: {
                        Image(systemName: isGridView ? "list.bullet" : "square.grid.2x2")
                    }

                    Button {
                        showFilterSheet = true
                    } label: {
                        Image(systemName: "slider.horizontal.3")
                    }
                }
                .padding(.horizontal)

                // Search Summary and Controls
                VStack(alignment: .leading, spacing: 8) {
                    HStack {
                        Text(isLoading ? "Searching..." : "About \(totalResults) results for \"\(searchText)\"")
                            .font(.caption)
                            .foregroundColor(.secondary)

                        Spacer()

                        Button {
                            showSortSheet = true
                        } label: {
                            HStack(spacing: 4) {
                                Image(systemName: "arrow.up.arrow.down")
                                    .font(.caption)
                                Text(sortBy.uppercased())
                                    .font(.caption)
                                    .fontWeight(.medium)
                            }
                        }
                        .foregroundColor(.accentColor)
                    }

                    // Active Filters
                    if !activeFilters.isEmpty {
                        ScrollView(.horizontal, showsIndicators: false) {
                            HStack(spacing: 8) {
                                ForEach(activeFilters, id: \.key) { filter in
                                    filterChip(filter.key, value: filter.value) {
                                        removeFilter(filter.key, value: filter.value)
                                    }
                                }
                            }
                            .padding(.horizontal)
                        }
                    }
                }
                .padding(.horizontal)
                .padding(.vertical, 8)
                .background(Color(.systemGray6))

                // Results
                if isLoading && results.isEmpty {
                    Spacer()
                    ProgressView()
                    Spacer()
                } else if results.isEmpty {
                    emptyStateView
                } else {
                    if isGridView {
                        gridView
                    } else {
                        listView
                    }
                }
            }
            .navigationBarHidden(true)
        }
        .onAppear {
            performSearch()
        }
        .sheet(isPresented: $showFilterSheet) {
            FilterScreen(initialFilters: currentFilters)
        }
        .sheet(isPresented: $showSortSheet) {
            SortScreen()
        }
    }

    private var activeFilters: [(key: String, value: String)] {
        var filters: [(key: String, value: String)] = []

        currentFilters.forEach { key, value in
            if let stringValue = value as? String, !stringValue.isEmpty {
                filters.append((key: key, value: stringValue))
            } else if let arrayValue = value as? [String], !arrayValue.isEmpty {
                arrayValue.forEach { item in
                    filters.append((key: key, value: item))
                }
            }
        }

        return filters
    }

    private var emptyStateView: some View {
        VStack(spacing: 16) {
            Spacer()

            Image(systemName: "magnifyingglass.circle")
                .font(.system(size: 64))
                .foregroundColor(.gray)

            Text("No results found")
                .font(.title2)
                .fontWeight(.bold)

            Text("Try adjusting your search or filters")
                .font(.body)
                .foregroundColor(.secondary)

            Button("Clear Filters") {
                currentFilters.removeAll()
                performSearch()
            }
            .buttonStyle(.borderedProminent)

            Spacer()
        }
    }

    private var listView: some View {
        ScrollView {
            LazyVStack(spacing: 16) {
                ForEach(results, id: \.id) { result in
                    resultListItem(result)
                }

                if hasMoreResults {
                    ProgressView()
                        .onAppear {
                            loadMoreResults()
                        }
                }
            }
            .padding()
        }
    }

    private var gridView: some View {
        ScrollView {
            LazyVGrid(columns: Array(repeating: GridItem(.flexible(), spacing: 16), count: 2), spacing: 16) {
                ForEach(results, id: \.id) { result in
                    resultGridItem(result)
                }

                if hasMoreResults {
                    GridRow {
                        ProgressView()
                            .gridCellColumns(2)
                            .onAppear {
                                loadMoreResults()
                            }
                    }
                }
            }
            .padding()
        }
    }

    private func resultListItem(_ result: SearchResult) -> some View {
        HStack(alignment: .top, spacing: 12) {
            // Image
            AsyncImage(url: URL(string: result.imageUrl ?? "")) { image in
                image
                    .resizable()
                    .aspectRatio(contentMode: .fill)
            } placeholder: {
                RoundedRectangle(cornerRadius: 8)
                    .fill(Color(.systemGray6))
                    .overlay {
                        Image(systemName: "photo")
                            .foregroundColor(.gray)
                    }
            }
            .frame(width: 80, height: 80)
            .clipShape(RoundedRectangle(cornerRadius: 8))

            VStack(alignment: .leading, spacing: 8) {
                // Title
                Text(highlightedText(result.title, query: searchText))
                    .font(.headline)
                    .fontWeight(.bold)
                    .lineLimit(2)

                // Description
                Text(result.description)
                    .font(.body)
                    .foregroundColor(.secondary)
                    .lineLimit(3)

                // Metadata
                HStack {
                    if let rating = result.rating {
                        HStack(spacing: 4) {
                            Image(systemName: "star.fill")
                                .foregroundColor(.yellow)
                                .font(.caption)
                            Text(String(format: "%.1f", rating))
                                .font(.caption)
                        }
                    }

                    if let category = result.category {
                        Text(category)
                            .font(.caption)
                            .padding(.horizontal, 8)
                            .padding(.vertical, 4)
                            .background(Color.accentColor.opacity(0.1))
                            .foregroundColor(.accentColor)
                            .cornerRadius(4)
                    }

                    Spacer()

                    if let price = result.price {
                        Text("$\(String(format: "%.2f", price))")
                            .font(.headline)
                            .fontWeight(.bold)
                            .foregroundColor(.accentColor)
                    }
                }
            }
        }
        .padding()
        .background(Color(.systemBackground))
        .cornerRadius(12)
        .shadow(color: .black.opacity(0.1), radius: 2, x: 0, y: 1)
    }

    private func resultGridItem(_ result: SearchResult) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            // Image
            AsyncImage(url: URL(string: result.imageUrl ?? "")) { image in
                image
                    .resizable()
                    .aspectRatio(contentMode: .fill)
            } placeholder: {
                RoundedRectangle(cornerRadius: 8)
                    .fill(Color(.systemGray6))
                    .overlay {
                        Image(systemName: "photo")
                            .foregroundColor(.gray)
                    }
            }
            .frame(height: 120)
            .clipShape(RoundedRectangle(cornerRadius: 8))

            VStack(alignment: .leading, spacing: 4) {
                Text(highlightedText(result.title, query: searchText))
                    .font(.subheadline)
                    .fontWeight(.bold)
                    .lineLimit(2)

                HStack {
                    if let rating = result.rating {
                        HStack(spacing: 2) {
                            Image(systemName: "star.fill")
                                .foregroundColor(.yellow)
                                .font(.caption2)
                            Text(String(format: "%.1f", rating))
                                .font(.caption2)
                        }
                    }

                    Spacer()

                    if let price = result.price {
                        Text("$\(Int(price))")
                            .font(.subheadline)
                            .fontWeight(.bold)
                            .foregroundColor(.accentColor)
                    }
                }
            }
        }
        .padding()
        .background(Color(.systemBackground))
        .cornerRadius(12)
        .shadow(color: .black.opacity(0.1), radius: 2, x: 0, y: 1)
    }

    private func filterChip(_ key: String, value: String, onRemove: @escaping () -> Void) -> some View {
        HStack(spacing: 4) {
            Text("\(key): \(value)")
                .font(.caption)

            Button {
                onRemove()
            } label: {
                Image(systemName: "xmark")
                    .font(.caption2)
            }
        }
        .padding(.horizontal, 8)
        .padding(.vertical, 4)
        .background(Color.accentColor)
        .foregroundColor(.white)
        .cornerRadius(12)
    }

    private func highlightedText(_ text: String, query: String) -> AttributedString {
        var attributedString = AttributedString(text)

        if !query.isEmpty {
            let ranges = text.ranges(of: query, options: .caseInsensitive)
            for range in ranges {
                let startIndex = attributedString.characters.index(
                    attributedString.startIndex,
                    offsetBy: text.distance(from: text.startIndex, to: range.lowerBound)
                )
                let endIndex = attributedString.characters.index(
                    attributedString.startIndex,
                    offsetBy: text.distance(from: text.startIndex, to: range.upperBound)
                )

                attributedString[startIndex..<endIndex].foregroundColor = .accentColor
                attributedString[startIndex..<endIndex].font = .bold
            }
        }

        return attributedString
    }

    private func performSearch(isNewSearch: Bool = true) {
        if isNewSearch {
            results.removeAll()
            currentPage = 1
            isLoading = true
            hasMoreResults = true
        }

        // Simulate API call
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.8) {
            let newResults = generateMockResults()
            if isNewSearch {
                results = newResults
                totalResults = 157 // Mock total
            } else {
                results.append(contentsOf: newResults)
            }
            isLoading = false
            isLoadingMore = false
            hasMoreResults = results.count < totalResults
        }
    }

    private func loadMoreResults() {
        guard !isLoadingMore && hasMoreResults else { return }

        isLoadingMore = true
        currentPage += 1
        performSearch(isNewSearch: false)
    }

    private func generateMockResults() -> [SearchResult] {
        return (0..<20).map { index in
            let baseIndex = (currentPage - 1) * pageSize + index
            return SearchResult(
                id: "result_\(baseIndex)",
                title: "Search Result \(baseIndex + 1) - \(searchText)",
                description: "This is a detailed description for search result \(baseIndex + 1). It contains relevant information about \(searchText) and provides useful context for the user.",
                imageUrl: "https://picsum.photos/300/200?random=\(baseIndex)",
                price: (baseIndex % 5 == 0) ? 29.99 + Double(baseIndex) * 2.5 : nil,
                rating: 3.0 + Double(baseIndex % 3),
                reviewCount: 50 + (baseIndex * 12),
                category: ["Electronics", "Books", "Clothing", "Home"][baseIndex % 4],
                tags: ["Popular", "New", "Featured"],
                dateAdded: Date().addingTimeInterval(-Double(baseIndex) * 86400),
                author: "Author \(baseIndex % 10)",
                source: "Source \(baseIndex % 5)"
            )
        }
    }

    private func removeFilter(_ key: String, value: String) {
        if var array = currentFilters[key] as? [String] {
            array.removeAll { $0 == value }
            if array.isEmpty {
                currentFilters.removeValue(forKey: key)
            } else {
                currentFilters[key] = array
            }
        } else {
            currentFilters.removeValue(forKey: key)
        }
        performSearch()
    }
}

// Helper extension for string ranges
extension String {
    func ranges(of substring: String, options: CompareOptions = []) -> [Range<Index>] {
        var ranges: [Range<Index>] = []
        var start = startIndex

        while let range = range(of: substring, options: options, range: start..<endIndex) {
            ranges.append(range)
            start = range.upperBound
        }

        return ranges
    }
}

#Preview {
    SearchResultsScreen(query: "Swift programming")
}